lisdf/components/control.py

```diff
@@ -0,0 +1,108 @@
+#! /usr/bin/env python3
+# -*- coding: utf-8 -*-
+# File   : control.py
+# Author : Jiayuan Mao
+# Email  : maojiayuan@gmail.com
+# Date   : 03/23/2022
+#
+# This file is part of lisdf.
+# Distributed under terms of the MIT license.
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from lisdf.utils.typing import Vector2f, Vector3f
+
+from .base import StringConfigurable
+
+__all__ = [
+    "ControlInfo",
+    "JointInfo",
+    "FixedJointInfo",
+    "HingeJointInfo",
+    "PrismaticJointInfo",
+]
+
+
+@dataclass
+class ControlInfo(object):
+    limited: bool
+    range: Vector2f
+
+    def __init__(self, limited: str, range: Vector2f):
+        assert limited in ("true", "false")
+        self.limited = limited == "true"
+        self.range = range
+
+
+@dataclass
+class JointInfo(StringConfigurable):
+    type_mapping = dict()  # type: ignore
+
+    def __init_subclass__(cls, type: str, **kwargs):
+        super().__init_subclass__(**kwargs)
+        setattr(cls, "type", type)
```

Comment:
Can you make this a property that each subclass must implement? Otherwise it's not clear that this is available

Reply:
But that will trigger issues for inheritance... Is there a way to declare a variable in dataclass that won't appear in the constructor?

Reply:
Can you make it an abstract property that needs to be implemented by subclasses? It returns a static value which is the type?

Like this: https://stackoverflow.com/a/48710068/6088502

Reply:
seems that abstractmethod can not work with dataclass. It will trigger errors.

lisdf/components/sensor.py:18: error: Only concrete class can be given where "Type[Sensor]" is expected

I added a property with NotImplmenetedError

Also left a TODO.

Reply:
That's weird! I just tested this and it works. Maybe some of the superclasses messing things up. It isn't very clean with the `__init_subclass__` but let's not worry too much about these details for now.

```
from abc import abstractmethod, ABC
from dataclasses import dataclass


@dataclass
class AbstractInfo(ABC):
    my_info: str

    @property
    @abstractmethod
    def type(self) -> str:
        raise NotImplementedError


class ConcreteInfo(AbstractInfo):
    @property
    def type(self) -> str:
        return "ConcreteInfo"


if __name__ == "__main__":
    info = ConcreteInfo(my_info="Concrete info")
    print(info.type)
    print(info.my_info)
```

Reply:
Let me check again. Maybe I did something wrong

Reply:
Ah I think it's because of the weird way I used to set the value (I was using setattr(cls, 'type')

There's a much simple/elegant way to set that! Just declare that variable as a ClassVar! (idk why I missed that...)